### ISSUE TITLE
tf.function breaks the gradient tape when looping over datasets

### ISSUE THREAD

**Issue by fachu000** [tensorflow/tensorflow#41594]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): uname -a:
Darwin Daniels-MacBook-Pro.local 18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64

- TensorFlow installed from (source or binary): binary

- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0

- Python version: Python 3.6.8 :: Anaconda, Inc.

Same bug in tf-nightly.

**Describe the current behavior**

The gradient tape is broken in graph mode when a statement is introduced in a loop. 

It works correctly in Eager mode or when the statement is taken out of the loop. 

**Describe the expected behavior**

As detailed in the documentation, a tf.function should either behave as in Eager mode or return an error. Here, none of those happen.

**Standalone code to reproduce the issue**

See [Colab notebook](https://colab.research.google.com/drive/1y__myjOuJEMF4izbJuoLbGqgDItDm7jn?usp=sharing).

```Python
import tensorflow as tf

@tf.function
def training_step(data, par):
    loss = tf.constant(0, dtype=tf.double)
    with tf.GradientTape() as tape:
        for d in data:
            loss = par # works if we take this line out of the loop
    return tape.gradient(loss, par)
    
par = tf.Variable(tf.zeros(1, dtype=tf.double))
data = tf.data.Dataset.from_tensor_slices([1,2,3])

tf.config.experimental_run_functions_eagerly(False)
print(f"graph mode: {training_step(data, par)}")

tf.config.experimental_run_functions_eagerly(True)
print(f"eager mode: {training_step(data, par)}")
  
```

Obviously, the same happens if `loss = par` is replaced with `loss += par`, as will typically be the case in a training loop.

**Other info / logs** 
The output is:

```
graph mode: None
eager mode: [1.]
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2, TF v2.3.0-rc2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/033e6d33d0099320d6ac1e56542b7f2f/41594-tf-nightly.ipynb). Thanks!

**Comment by sushreebarsa** [comment]
Was able to reproduce this issue in TF 2.6.0-dev20210526 ,please find the gist[ here ](https://colab.research.google.com/gist/sushreebarsa/8364f9b32e8dfcb394431015be5a4411/untitled9.ipynb)..Thanks !

**Comment by synandi** [comment]
@fachu000, Kindly use `[tf.data.Iterator](https://www.tensorflow.org/api_docs/python/tf/data/Iterator)` API, which is the primary mechanism for enumerating elements of a [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset). AutoGraph it translates into a `tf.while_loop` . Please find the gist of working code [here](https://colab.sandbox.google.com/gist/synandi/01c22c6693e3576b248a5e1577e99f1c/41594.ipynb). Thank you!

**Comment by tilakrayal** [comment]
@fachu000,
I tried to execute the mentioned code with the alternative approach by using **loss = tf.constant((0,), dtype=tf.double)**
 and it was executed without any issue/error with the intended output. Kindly find the [gist](https://colab.research.google.com/gist/tilakrayal/b494f5f3cfad8a214b51d74c9bfd34d9/untitled1560.ipynb) of the work-around. Thank you!
